### ISSUE TITLE
Lifecycle hook to properly drain instances

### ISSUE THREAD

**Issue by nitrocode** [spacelift-io/terraform-aws-spacelift-workerpool-on-ec2#25]
It would be handy to have a lifecycle hook on the autoscaling group to safely spin down instances when updating the AMI. The hook could pause the ec2 termination, wait for spacelift to finish any remaining jobs on its agents, drain the instance, and continue the termination.

## references

* https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/autoscaling_lifecycle_hook
* https://circleci.com/blog/graceful-shutdown-using-aws/
* https://github.com/spacelift-io/terraform-aws-spacelift-workerpool-on-ec2/pull/22

**Comment by xophere** [comment]
Yes please.

**Comment by lorengordon** [comment]
Yeah, this would be very handy. I have a spacelift stack that is actually managing this worker pool, and the pool is linked to the stack. A little self-referential I guess. Which is of course the problem. But, as it is, when I modify the launch template, the instance refresh kicks off, which terminates the runner that the stack run is on! Some kind of lifecycle hook to properly drain instances would resolve the problem without complicating the architecture.

**Comment by lorengordon** [comment]
Resolution may be imminent... https://github.com/spacelift-io/terraform-aws-spacelift-workerpool-on-ec2/pull/150

**Comment by zhimsel** [comment]
@Apollorion this can be closed
